packages/selection/src/hooks/useKeyboardFocus.js

```diff
@@ -0,0 +1,52 @@
+/**
+ * Copyright Zendesk, Inc.
+ *
+ * Use of this source code is governed under the Apache License, Version 2.0
+ * found at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+import { useState } from 'react';
+
+import composeEventHandlers from '../utils/composeEventHandlers';
+
+export default function useKeyboardFocus() {
+  const [keyboardFocused, setKeyboardFocused] = useState(false);
+  let keyboardFocusable = true;
+
+  const onMouseDown = () => {
+    keyboardFocusable = false;
+
+    /**
+     * This is necessary to recognize focus events caused by keyboard vs mouseDown.
+     * Due to React event ordering this is always called before onFocus.
+     */
+    setTimeout(() => {
+      keyboardFocusable = true;
+    }, 0);
+  };
+
+  const onFocus = () => {
+    if (keyboardFocusable) {
+      setKeyboardFocused(true);
+    }
+  };
+
+  const onBlur = () => {
+    setKeyboardFocused(false);
+  };
+
+  const getFocusProps = ({ tabIndex = 0, ...props } = {}) => {
+    return {
+      tabIndex,
+      onBlur: composeEventHandlers(props.onBlur, onBlur),
+      onFocus: composeEventHandlers(props.onFocus, onFocus),
+      onMouseDown: composeEventHandlers(props.onMouseDown, onMouseDown),
```

Comment:
Is there more we should be doing here in order to cancel keyboard modality? See `onPointerDown` composition under https://github.com/WICG/focus-visible/blob/master/src/focus-visible.js for example.

Reply:
Would we feel comfortable breaking out the non-selection items from this PR?

If we are doing full-rewrites of these things it may be easier to review them independently.

Reply:
Fine with me. Can someone add a new issue to our backlog?

Reply:
yep I think that makes sense I'll create the issue now

Reply:
I'll remove them from this PR and we might need to evaluate whether we need to move the util methods out of here to.